### DISCUSSION TITLE
SkiaSharp and DirectX?

### DISCUSSION THREAD

**Discussion by bforl** [mono/SkiaSharp#2911] [General]
From my understanding, SKIA has lots of different backends for rendering, The default being a CPU renderer, but then hardware renderers for different platforms, OpenGL, DirectX etc

I see that for OpenGL, there is an SKGLElement, but how do you use DirectX?

In my case, I am using WPF. I have tried the SKGLElement, but from my experiments, its often not that much quicker, and in some cases, it's actually quite a bit slower (DrawPoints seems to be a lot slower)?

**Comment by nockawa** [comment]
I'm a beginner on this lib as you, but I have the same requirement. From my current understanding, if you look at the [compatibility matrix](https://github.com/mono/SkiaSharp/wiki/Views-Support-Matrix) you'll see that WPF's renderer is software only, unless you use a WinForm Host Control, then you'll use OpenGL.

From what I saw in Avalonia's source code, there's a way to use the ANGLE library (which is an OpenGL to DX translator) to render Skia with a D3D target.

For now, and the use I have, I rely on WinForm, but I wonder if that is the fastest renderer on Windows for SkiaSharp, because I need to draw thousands of primitives.

**Reply by bforl** [reply]
I have been using the SKGLElement class, that allows you to use OpenGL in WPF without the WinForms host bit.

But oddly, it uses static classes, so you have issues with multiples instances and/or trying to use it multi threaded.

But even when you do use SKGLElement, it's really not that much quicker, and as mentioned above, can even be slower?

Maybe I am misunderstanding something, but it seems odd not to just render using DirectX when running SKIA in WPF (as its windows only)?

It's also odd that this is a Microsoft Project, and we only have the SKGLElement for hardware rendering which has the above issues.

Again, apologies if I am missing something, but it seems like a good amount of functionality is missing?

Is Avalonia using SKGLElement?

**Comment by nockawa** [comment]
My understanding is #1893 was made to add ANGLE support to WinUI, but nothing more to WPF.
You can indeed use SLGLElement (with the 3.0 pre-release package) to rely on OpenGL with WPF.
SkiaSharp is heavily cross-platform so it makes sense it relies on OpenGL, but yes, a D3D layer for WPF would be a nice thing.
Avalonia runs on many platforms, they use Vulkan, OGL and ANGLE apparently.

**Reply by bforl** [reply]
Yeah, I appreciate Skia is cross platform, but I thought they specifically added a Directx11 binding?

**Reply by nockawa** [reply]
What are you looking to achieve? Because if you need very low level, I suggest maybe to take a look at [Silk.net](https://github.com/dotnet/Silk.NET), I'm sure they have a D11/Skia integration. How would it fit into WPF, that I don't know.

**Reply by bforl** [reply]
> What are you looking to achieve?

Just better performance really, as well as hopefully a solution to the multiple instance/threading issue.

**Reply by nockawa** [reply]
I'm currently using a Winforms base with OpenGL, perfs seems decent, I draw 16000 rect and present them in about 3ms.
If at some point this seems too slow for me, I think I would investigate how Avalonia did their renderer (which is based on Skia now).
A quick look [here](https://github.com/search?q=repo%3AAvaloniaUI%2FAvalonia%20ISkiaGPU&type=code) you can see they rely on Vulkan, OGL and Metal.
So my take: they use Vulkan on Windows, but it's pure speculation.

**Comment by ghost** [comment]
Not a huge fan of SkiaSharp in WPF for the reasons you stated.

SKGLElement also has limitations, I believe the gl context is synced to an underlying directx context, which makes multiple SKGLElements impractical, and can't be good for performance either.

Pretty sure I have a bug report open for this still.
